### PR TITLE
Add Vue core package

### PR DESCRIPTION
**Pull request by lorisleiva**
Hi there 👋 

This PR adds a new core package that supports Vue 3.

It works by providing two functions:
- `initWallet` that initialises the wallet store and should be called only once.
- `useWallet` that returns that store by reference so we can use its content anywhere in our application.

### Basic usage

```js
import { initWallet, useWallet } from '@solana/wallet-adapter-vue';
import { getPhantomWallet } from '@solana/wallet-adapter-wallets';

// Initialise the wallet store once.
const wallets = [getPhantomWallet()];
initWallet({ wallets, autoConnect: true });

// Use its content anywhere.
const { wallet, select, connect, disconnect, connected, sendTransaction } = useWallet();
select('Phantom');
connect();
sendTransaction(...);
```

### References

Analogously to the React core package, here are the options you can provide to the `initWallet` method.

```ts
interface WalletStoreProps {
    wallets: Wallet[];
    autoConnect?: boolean;                  // Default: false
    onError?: (error: WalletError) => void; // Default: (error: WalletError) => console.error(error)
    localStorageKey?: string;               // Default: "walletName"
}
```

And here is the definition of the wallet store you can access from anywhere in your Vue application once initialised.

```ts
interface WalletStore {
    // Props.
    wallets: Wallet[];
    autoConnect: boolean;

    // Data.
    walletName: Ref<string | null>,
    walletsByName: Ref<WalletDictionary>,
    wallet: Ref<Wallet | null>;
    adapter: Ref<Adapter | null>;
    publicKey: Ref<PublicKey | null>;
    ready: Ref<boolean>;
    connected: Ref<boolean>;
    connecting: Ref<boolean>;
    disconnecting: Ref<boolean>;

    // Methods.
    select(walletName: WalletName): void;
    connect(): Promise<void>;
    disconnect(): Promise<void>;
    sendTransaction(transaction: Transaction, connection: Connection, options?: SendTransactionOptions): Promise<TransactionSignature>;
    signTransaction(transaction: Transaction): Promise<Transaction>;
    signAllTransactions(transaction: Transaction[]): Promise<Transaction[]>;
    signMessage(message: Uint8Array): Promise<Uint8Array>;
}
```

### Why not provide/inject?
In React, it is common to use Context components and retrieve their content within their descendants. Whilst Vue has a similar concept via the `provide` and `inject` API, they aren't commonly used in the Vue community. Instead, it is more common to create a global store using the composition API and access its content from anywhere when needed, which is why I went for that approach in this PR.

### The `isolatedModules` flag
Note that I had to override the following TS option `"isolatedModules": false` inside the vue package in order to import from `@vue/runtime-core` (See https://github.com/solana-labs/wallet-adapter/issues/67#issuecomment-929386709).

### Manual testing
I manually tested this package locally [in my solana test app](https://github.com/lorisleiva/solana-vue3-counter/tree/feature/test-local-vue-adapter) by installing it as a symlink — i.e. using `"@solana/wallet-adapter-vue": "file:../../wallet-adapter/packages/core/vue"`. Everything worked properly and I was able to sign and send some transactions using my Phantom wallet.

However, I noticed that components using the references (`Ref`) provided by the store did not update when the references changed. I can see the references updating in real time in the Vue dev tools but for some reason the component itself does not re-render itself. When I call `this.$forceUpdate` on the component, I can see all the latest values being updated properly. I believe this is just a quirk from installing the package locally using a symlink but I thought I'd mention it in case this issue ring a bell to someone.

### CODE REVIEW

File: packages/core/vue/src/useWallet.ts [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+        if (! adapter.value) throw newError(new WalletNotSelectedError());
+        if (! connected.value) throw newError(new WalletNotConnectedError());
+        return await adapter.value.sendTransaction(transaction, connection, options);
+    };
+
+    // Sign a transaction if the wallet supports it.
+    const signTransaction = async (transaction: Transaction) => {
+        if (! adapter.value) throw newError(new WalletNotSelectedError());
+        if (! connected.value) throw newError(new WalletNotConnectedError());
+        if (!('signTransaction' in adapter.value)) throw newError(new OperationNotSupportedByWalletError());
```
Comment (jordansexton) [inline packages/core/vue/src/useWallet.ts:174]: Throwing a runtime error here is something the React API design intentionally avoids. The optional methods will be undefined if the wallet doesn't support them. This produces type safety for the developer that calls them and avoids an error the user sees.

File: packages/core/vue/src/errors.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import { WalletError } from '@solana/wallet-adapter-base';
+
+export class WalletNotSelectedError extends WalletError {
+    name = 'WalletNotSelectedError';
+}
+
+export class OperationNotSupportedByWalletError extends WalletError {
+    name = 'OperationNotSupportedByWalletError';
+}
```
Comment (jordansexton) [inline packages/core/vue/src/errors.ts:9]: This shouldn't be needed if the operations are typesafe.

File: packages/core/vue/src/useLocalStorage.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import { customRef, Ref } from '@vue/reactivity';
+
+export function useLocalStorage<T> (key: string, defaultValue: T | null = null): Ref<T | null> {
+    return customRef<T | null>((track, trigger) => ({
+        get: () => {
+            track();
+            const value = localStorage.getItem(key);
+            return value ? JSON.parse(value) : defaultValue;
```
Comment (jordansexton) [inline packages/core/vue/src/useLocalStorage.ts:8]: We don't do this yet in the React design, but we should -- wrap `JSON.parse` in try/catch and return the default value if it fails. This way users won't get an uncaught runtime error.

File: packages/core/vue/src/useLocalStorage.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+        },
+        set: value => {
+            if (value === null) {
+                localStorage.removeItem(key);
+            } else {
+                localStorage.setItem(key, JSON.stringify(value));
```
Comment (jordansexton) [inline packages/core/vue/src/useLocalStorage.ts:14]: Should probably do the same here, since `JSON.stringify` will throw for circular references.

File: packages/core/vue/src/useWallet.ts [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+type WalletDictionary = { [key: string]: Wallet };
+
+export interface WalletStore {
+    // Props.
+    wallets: Wallet[];
+    autoConnect: boolean;
+
+    // Data.
+    walletName: Ref<string | null>,
+    walletsByName: Ref<WalletDictionary>,
```
Comment (jordansexton) [inline packages/core/vue/src/useWallet.ts:23]: Let's continue to use the available `WalletName` enum here for type safety.

File: packages/core/vue/src/useWallet.ts [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+    localStorageKey = 'walletName',
+}: WalletStoreProps): void => {
+    const walletName = useLocalStorage<string>(localStorageKey);
+    const wallet = ref<Wallet | null>(null);
+    const adapter = ref<Adapter | null>(null);
+    const publicKey = ref<PublicKey | null>(null);
+    const ready = ref<boolean>(false);
+    const connected = ref<boolean>(false);
+    const connecting = ref<boolean>(false);
+    const disconnecting = ref<boolean>(false);
```
Comment (jordansexton) [inline packages/core/vue/src/useWallet.ts:66]: I'm not sure how this applies to Vue, but the way this state is handled has changed in the React package:
https://github.com/solana-labs/wallet-adapter/blob/581cb52a41ad7a255542de39115d051b0833e57c/packages/core/react/src/WalletProvider.tsx#L42

This allows all related state updates to happen at once, which prevents inconsistent state, rerenders, and race conditions. I see you are updating the state below in one place which is great.

It may be worth looking at how the `connecting`/`disconnecting`/`working` values are handled now also to avoid rerenders or `useEffect` invocations (again with the caveat that I don't know how relevant to Vue this is).

File: packages/core/vue/src/useWallet.ts [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+
+export const useWallet = (): WalletStore => walletStore;
+
+export const initWallet = ({
+    wallets,
+    autoConnect = false,
+    onError = (error: WalletError) => console.error(error),
+    localStorageKey = 'walletName',
+}: WalletStoreProps): void => {
+    const walletName = useLocalStorage<string>(localStorageKey);
```
Comment (jordansexton) [inline packages/core/vue/src/useWallet.ts:59]: Can't this be `null`?

File: packages/core/vue/src/useWallet.ts [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+    const walletsByName = computed<WalletDictionary>(() => {
+        return wallets.reduce((walletsByName, wallet) => {
+            walletsByName[wallet.name] = wallet;
+            return walletsByName;
+        }, {} as WalletDictionary);
+    });
+
+    // Update the wallet and adapter based on the wallet provider.
+    watch(walletName, (): void => {
+        wallet.value = walletsByName.value?.[walletName.value as string] ?? null;
```
Comment (jordansexton) [inline packages/core/vue/src/useWallet.ts:77]: What does `walletName.value as string` do? Isn't the type of `walletName` known?

File: packages/core/vue/src/useWallet.ts [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+        if (adapter.value) {
+            ready.value = adapter.value.ready;
+            publicKey.value = adapter.value.publicKey;
+            connected.value = adapter.value.connected;
+        } else {
+            ready.value = false;
+            publicKey.value = null;
+            connected.value = false;
+        }
+    }, { immediate:true });
```
Comment (jordansexton) [inline packages/core/vue/src/useWallet.ts:88]: Will you explain what this flag does for me?

File: packages/core/vue/src/useWallet.ts [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+        if (! adapter.value) return;
+        ready.value = adapter.value.ready;
+        publicKey.value = adapter.value.publicKey;
+        connected.value = adapter.value.connected;
+    };
+    watchEffect(onInvalidate => {
+        if (! adapter.value) return;
+        adapter.value.on('ready', onReady);
+        adapter.value.on('connect', onConnect);
+        adapter.value.on('error', onError);
```
Comment (jordansexton) [inline packages/core/vue/src/useWallet.ts:109]: What about `disconnect`?

File: packages/core/vue/src/useWallet.ts [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+        onInvalidate(() => {
+            if (! adapter.value) return;
+            adapter.value.off('ready', onReady);
+            adapter.value.off('connect', onConnect);
+            adapter.value.off('error', onError);
+        });
+    });
+
+    // Helper method to return an error whilst using the onError callback.
```
Comment (jordansexton) [inline packages/core/vue/src/useWallet.ts:118]: Smart! We should probably do this elsewhere.

File: packages/core/vue/src/useWallet.ts [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+        if (! connected.value) throw newError(new WalletNotConnectedError());
+        if (!('signMessage' in adapter.value)) throw newError(new OperationNotSupportedByWalletError());
+        return await adapter.value.signMessage(message);
+    };
+
+    // If autoConnect is enabled, try to connect when the adapter changes and is ready.
+    watchEffect(async (): Promise<void> => {
+        if (! autoConnect || ! adapter.value || ! ready.value || connected.value || connecting.value) return;
+        await connect();
+    });
```
Comment (jordansexton) [inline packages/core/vue/src/useWallet.ts:198]: The implementation has changed significantly in the React package:
https://github.com/solana-labs/wallet-adapter/blob/581cb52a41ad7a255542de39115d051b0833e57c/packages/core/react/src/WalletProvider.tsx#L70-L87

Several types of bugs were eliminated. This also shouldn't call the `connect` method because that can open a popup if the wallet isn't available.

File: packages/core/vue/src/useWallet.ts [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+
+    walletStore = {
+        // Props.
+        wallets,
+        autoConnect,
+
+        // Data.
+        walletName,
+        walletsByName,
```
Comment (jordansexton) [inline packages/core/vue/src/useWallet.ts:207]: These shouldn't need to be exposed.

**Review by jordansexton — COMMENTED** [review]
Overall looks good so far! I'll try to run it in your example package with a UI to see how it feels.

**Comment by lorisleiva** [comment]
Thanks Jordan! I'll go through each of these points gradually. 🙂

File: packages/core/vue/src/useWallet.ts (hunk 9 above)
Comment (lorisleiva) [reply packages/core/vue/src/useWallet.ts:88]: In Vue:

- `watchEffect(fn)` is the equivalent of `useEffect(fn)` but you cannot provide a list of variables to listen for, instead it will automatically recognise any reactive variables inside the function and call that method every time one or more of these variables change during a given tick. Most of the time, this is exactly what we need and so we let Vue handle the list of vars for us.
- If you want to explicitly tell Vue what you're watching, you can use `watch(var, fn)` or `watch([vars], fn)`.
- However, by default, the `watch` method will only be called on future updates and not immediately after defining it (as `watchEffect` and `useEffect` do).
- Therefore, you need to provide that flag to tell it to trigger immediately.
- In conclusion, `watch([vars], fn, { immediate:true })` is equivalent to `useEffect(fn, [vars])`.

File: packages/core/vue/src/useWallet.ts (hunk 10 above)
Comment (lorisleiva) [reply packages/core/vue/src/useWallet.ts:109]: I had issues with the `onDisconnect` callback when defined as `() => walletName.value = null`.

When using the `autoconnect` flag, it would cause the whole thing to disconnect on refresh.

The flow was:
- Connect your wallet using the `autoconnect` flag.
- Refreshing the page would call `onDisconnect` on the adapter.
- On reload, the walletName would be empty and so we wouldn't be able to autoconnect.

Additionally, we explicitly set `walletName.value = null` in the `finally` scope of the `disconnect` method so I thought this was perhaps redundant.

I'm happy to give it another go though, did you come across that issue with React?

File: packages/core/vue/src/useWallet.ts (hunk 7 above)
Comment (lorisleiva) [reply packages/core/vue/src/useWallet.ts:59]: It can, the `null` value is defined in the return type of the signature:

```ts
useLocalStorage<T> (key: string, defaultValue: T | null = null): Ref<T | null>;
```

I wasn't sure how to make this more explicit whilst making sure we allow for null values in the signature.

File: packages/core/vue/src/useWallet.ts (hunk 8 above)
Comment (lorisleiva) [reply packages/core/vue/src/useWallet.ts:77]: The type of this variable

```ts
const walletName = useLocalStorage<string>(localStorageKey);
```

Resolves to `Ref<string | null>` so `walletName.value` resolves to `string | null` and TS was complaining.

I should probably mark the `walletName` explicitly like 


```ts
const walletName:Ref<string | null> = useLocalStorage<string>(localStorageKey);
```

File: packages/core/vue/src/useWallet.ts (hunk 6 above)
Comment (lorisleiva) [reply packages/core/vue/src/useWallet.ts:66]: I can do something similar but I don't think this will change anything in Vue because the code will be executed within the same DOM update cycle (a.k.a a tick).

But I can of course group it in a state object so that it is at least consistent with the logic of the other packages.

---

On a similar note, I had issues with reactivity if I didn't provide the `Refs` directly in the `WalletStore` but I'll try again to give the real values instead of the `Ref` variables.

To give more context, in Vue `Refs` are equivalent to `useState` but the API is quite different:

```js
// In React.
const [foo, setFoo] = useState('defaultFoo'); 
foo; // Access
setFoo('newFoo'); // Edit

// In Vue.
const foo = ref('defaultFoo');
foo.value; // Access
foo.value = 'newFoo'; // Edit
```

So ultimately, by returning the `Ref` we are also giving the equivalent of the `setFoo` method which I don't think is what we want. I'll see what I can do. 🙂

Comment (lorisleiva) [reply packages/core/vue/src/useWallet.ts:66]: Okay I've done some tests and If I don't return `Refs` reactivity simply does not work in Vue for primitive values.

An alternative would be to wrap the entire store in the `reactive` function, but that would still allow us to update any value inside it (including functions) and would differ from what most people are doing when creating Vue 3 libraries.

Also, the nice thing about returning `Refs` is that they are perfectly understood by Vue's components and their template knows how to handle them even without using `.value`.

File: packages/core/vue/src/useWallet.ts (hunk 7 above)
Comment (lorisleiva) [reply packages/core/vue/src/useWallet.ts]: I've explicitly defined it as:

```js
const walletName: Ref<WalletName | null> = useLocalStorage<WalletName>(localStorageKey);
```

File: packages/core/vue/src/useLocalStorage.ts (hunk 4 above)
Comment (lorisleiva) [reply packages/core/vue/src/useLocalStorage.ts]: All done. Am I right to assume we want to fail silently if that happens?

Btw, there is already an implementation of `useLocalStorage` in the `@vueuse/core` npm package but I assumed we didn't want an extra dependency just for that. Let me know, if you'd rather I use it instead.

https://vueuse.org/core/useLocalStorage/

Comment (jordansexton) [reply packages/core/vue/src/useLocalStorage.ts]: I think it's okay to just fail to add the item. We might want to `console.error` though, wdyt?

File: packages/core/vue/src/useWallet.ts (hunk 6 above)
Comment (jordansexton) [reply packages/core/vue/src/useWallet.ts:64]: Ah, okay. The behavior you had doesn't necessarily need to be changed, as long as nothing can react to those individual value changes before they are all changed.

File: packages/core/vue/src/useWallet.ts (hunk 7 above)
Comment (jordansexton) [reply packages/core/vue/src/useWallet.ts]: Can you help me understand what a Ref does in Vue?

File: packages/core/vue/src/useWallet.ts (hunk 10 above)
Comment (jordansexton) [reply packages/core/vue/src/useWallet.ts]: > Refreshing the page would call onDisconnect on the adapter.

This one is interesting, and seems likely to be the cause. I wonder why that is, as I don't think React does this?

Comment (jordansexton) [reply packages/core/vue/src/useWallet.ts]: A _possible_ workaround could be to add an `onbeforeunload` event listener to unregister the `disconnect` event handler so this doesn't happen. But we do need a `disconnect` event listener, since adapters can disconnect for various reasons outside the control of the app.

File: packages/core/vue/src/useLocalStorage.ts [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+
+export function useLocalStorage<T> (key: string, defaultValue: T | null = null): Ref<T | null> {
+    return customRef<T | null>((track, trigger) => ({
+        get: () => {
+            track();
+            const value = localStorage.getItem(key);
+            try {
+                return value ? JSON.parse(value) : defaultValue;
+            } catch (error) {
+                return defaultValue;
```
Comment (jordansexton) [inline packages/core/vue/src/useLocalStorage.ts:11]: I'm thinking we should `console.warn` here, and `console.error` in the setter

File: packages/core/vue/src/useWallet.ts [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+    WalletNotConnectedError,
+    WalletNotReadyError,
+    WalletError,
+} from '@solana/wallet-adapter-base';
+import { Wallet, WalletName } from '@solana/wallet-adapter-wallets';
+import { useLocalStorage } from './useLocalStorage';
+import { WalletNotSelectedError } from './errors';
+
+type Adapter = ReturnType<Wallet['adapter']>;
+type WalletDictionary = { [key: string]: Wallet };
```
Comment (jordansexton) [inline packages/core/vue/src/useWallet.ts:14]: The key here should be a `[name in WalletName]` if possible

File: packages/core/vue/src/useWallet.ts [hunk 16]
```diff
@@ -0,0 +1,264 @@
...
+    // Send a transaction using the provided connection.
+    const sendTransaction = async (transaction: Transaction, connection: Connection, options?: SendTransactionOptions) => {
+        if (! adapter.value) throw newError(new WalletNotSelectedError());
+        if (! connected.value) throw newError(new WalletNotConnectedError());
+        return await adapter.value.sendTransaction(transaction, connection, options);
+    };
+
+    // Sign a transaction if the wallet supports it.
+    const signTransaction = computed(() => {
+        if (! (adapter.value && 'signTransaction' in adapter.value)) return undefined;
```
Comment (jordansexton) [inline packages/core/vue/src/useWallet.ts:198]: Nice, this is pretty elegant

File: packages/core/vue/src/useWallet.ts [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+        }
+    });
+
+    // Sign multiple transactions if the wallet supports it
+    const signAllTransactions = computed(() => {
+        if (! (adapter.value && 'signAllTransactions' in adapter.value)) return undefined;
+        return async (transactions: Transaction[]) => {
+            if (! connected.value) throw newError(new WalletNotConnectedError());
+            // @ts-ignore
+            return await adapter.value.signAllTransactions(transactions);
```
Comment (jordansexton) [inline packages/core/vue/src/useWallet.ts:212]: Would `adapter.value.signAllTransactions!` work instead of ts-ignore?

File: packages/core/vue/src/useWallet.ts (hunk 9 above)
Comment (jordansexton) [reply packages/core/vue/src/useWallet.ts:113]: Thanks for the detailed explanation!

**Review by jordansexton — COMMENTED** [review]
Looks great, this seems nearly ready to merge

File: packages/core/vue/src/useLocalStorage.ts (hunk 4 above)
Comment (lorisleiva) [reply packages/core/vue/src/useLocalStorage.ts]: Good shout! 👍

File: packages/core/vue/src/useWallet.ts (hunk 7 above)
Comment (lorisleiva) [reply packages/core/vue/src/useWallet.ts]: Of course! The following code in React:

```js
const [foo, setFoo] = useState('defaultFoo');
foo; // Access
setFoo('newFoo'); // Edit
```

Can be achieved in Vue this way:

```js
const foo = ref('defaultFoo')
foo.value // Access
foo.value = 'newFoo'; // Edit
```

This ensures that reactivity is kept even when dealing with primitive values as they are wrapped in a reactive object.

Alternatively, one can use the `reactive` method provided by Vue to make an object and its properties reactive. However, if you then destructure that object, reactivity will be lost on primitive properties.

```js
const state = reactive({
    foo: 'defaultFoo',
    bar: { some: 'object' },
})
state.foo // Access
state.foo = 'newFoo'; // Edit

const { foo, bar } = state;
// `foo` is no longer reactive
// `bar` is now a `Reactive` object

state.foo = 'newFoo -- edited after destructuring';
state.bar.some = 'object -- edited after destructuring';

console.log(foo); // Outputs: newFoo
console.log(bar); // Outputs: { some: 'object -- edited after destructuring' }
```

When dealing with a `reactive` object, you can also destructure all of its properties into `Refs` using the `toRefs` method. This will ensure even primitive properties are reactive but they will be wrapped in a `Ref` object and so you'll need to access them using the `.value` attribute.

```js
const state = reactive({
    foo: 'defaultFoo',
    bar: { some: 'object' },
})

const { foo, bar } = toRefs(state);
state.foo = 'newFoo -- edited after destructuring';

console.log(foo.value); // Outputs: newFoo -- edited after destructuring
```

So in the end, when dealing with primitive values (or any values to stay consistent within our code), using `Refs` in Vue is the go-to approach. Additionally, when handling `Refs` inside templates (the equivalent of JSX in Vue), Vue will properly recognised `Refs` variables so you can access and edit them directly without using the `.value` attribute.

File: packages/core/vue/src/useWallet.ts (hunk 10 above)
Comment (lorisleiva) [reply packages/core/vue/src/useWallet.ts]: It worked! I understand what was happening now.

Just like in React, the `onInvalidate` callback (that disables all adapter listeners) defined in the `watchEffect` method will be called when the component is unmounted. From the [documentation](https://v3.vuejs.org/guide/reactivity-computed-watchers.html#side-effect-invalidation):

> This invalidation callback is called when:
>
> - the effect is about to re-run
> - the watcher is stopped (i.e. when the component is unmounted if watchEffect is used inside setup() or lifecycle hooks)

However, in the `useWallet.ts` file, we're not inside a particular component as we're defining a global store. In React, it is more common to define these in `ContextProviders` which are components themselves giving data to their descendants. In Vue, we have something similar with `provide` and `inject` but it's not quite the same and is rarely used in the community.

So we can do two things to go around this:
- Since a component will ultimately be calling the `initWallet` method, we could ask for a component as a parameter and then give it to the `watchEffect` method so it hooks itself to whenever that component unmounts. The only issue with this approach is we're dirtying a little bit the API by having to ask the component to add itself to the parameters.
- Do what you suggested and add a `beforeunload` window listener that invalidates the adapter listeners.

Both of these make sure adapter listeners are invalidated before the `onDisconnect` callback is triggered by the wallet itself upon refresh.

I went for the second approach to keep the API cleaner and less coupled to components. Do let me know if you'd rather I went for the first approach (or any other approach).

Thanks for your help in this it was really bugging me and I feel much better now that I understand what's happening haha.

**Comment by lorisleiva** [comment]
Thanks! I've been through all your points. Almost there! 💪 😊

**Comment by Bigandy4climate** [comment]
Hey @lorisleiva, it says your package is not yet published on npm :(

File: packages/core/vue/src/useWallet.ts (hunk 10 above)
Comment (jordansexton) [reply packages/core/vue/src/useWallet.ts]: My concern about the `beforeunload` event is that I don't know if it could result in a race condition. For example, if the wallet's `disconnect` is triggered by `beforeunload` or something, and that event handler is called before the one here. I'm not saying we shouldn't do this, but it's something to be aware of, and could be nasty to debug if reported by users.

Comment (jordansexton) [reply packages/core/vue/src/useWallet.ts]: We could maybe add a `cleanup`/`teardown` method to the store, and then a top level component unmounting could call it? It's guaranteed that devs will forget to do this though...

File: packages/core/vue/src/useWallet.ts (hunk 7 above)
Comment (jordansexton) [reply packages/core/vue/src/useWallet.ts]: Thanks a lot, that makes perfect thanks.

Comment (jordansexton) [reply packages/core/vue/src/useWallet.ts]: It seems like the way Vue works is kinda like MobX then. It tracks access of reactive values when they are dereferenced by getting/setting `ref.value`, so it knows what functions to call and what component trees to render.

**Comment by Bigandy4climate** [comment]
Hey unsure where to post this comment but I got 'unexpected token "export"' upon running this line :(


import { initWallet, useWallet } from '@solana/wallet-adapter-vue'

**Comment by Bigandy4climate** [comment]
Oh yeah I am on Nuxt :( Is that relevant info?

**Comment by lorisleiva** [comment]
I've not played with Nuxt before but yeah I think you might need to tweak your webpack settings.

**Comment by Bigandy4climate** [comment]
In tsconfig, target was ES2018, switched it to ESNext with no result :cry:

**Comment by lorisleiva** [comment]
Sorry I'm not sure I'm very qualified to help here I struggle so much when configuring bundlers/compilers.

All I can say is `unexpected token "export"` usually means your compiler wants modules to use `module.export = {}` and does not accept modules that export this way:

```js
export default { foo: 'bar' };
export const myAdditionalVariable = 42;
```

What I usually do when struggling with configuring compilers is start with a brand new application and only add the dependency that's causing me an issue to understand what I need to change to make it work. Then I bring my findings to my current application. I hope this helps somehow. 🍀

**Comment by Bigandy4climate** [comment]
https://stackoverflow.com/a/69516471/10620200

**Comment by Bigandy4climate** [comment]
This is weird actually because Nuxt does support ESModules in the src so why not node modules too?

**Comment by Bigandy4climate** [comment]
Hi @lorisleiva, 

What does the select method do? 
Also, intellisense for type Wallet is very inconsistent on my install (randomly changing), while I get errors claiming that Wallet objects are self-referring objects 😕

**Comment by lorisleiva** [comment]
Hi 👋 `select` selects a type of wallet (such as Phantom) by providing a `WalletName` (e.g. `Phantom`).

I'm not sure about the intellisense issues though.
